### DISCUSSION TITLE
Link between invoice and invoiced timesheets/expenses

### DISCUSSION THREAD

**Discussion by kernie** [kimai/kimai#5166] [Questions & Answers]
Hi,

I was testing the Kimai API to check whether I can get all needed information of an invoice to create the XML file of the electronic invoice (XRechnung, ZUGFERD, see request #1396).

However, seeing the [API definition](https://demo.kimai.org/api/doc) and reading #4738 it seems that there isn't an obvious link between the created invoice and the invoiced items.

If this observation is correct, then
- what happens, when I cancel an invoice? Are the former invoiced items reset to 'unexported'?
- how can I make sure, that I invoice all items, even if somebody exported timesheets beforehand (e.g. to show customer)
- (my current struggle) how can I get the invoiced timesheets of a created invoice to process outside Kimai?

Thanks!

**Comment by kevinpapst** [comment]
If you want to generate the invoice, use one of the invoice events.
But I am not going to help on this topic, as I will built the Zugferd support anyway. 

> what happens, when I cancel an invoice? Are the former invoiced items reset to 'unexported'?

No.

> how can I make sure, that I invoice all items, even if somebody exported timesheets beforehand (e.g. to show customer)

It's up to you to come up with a workflow.

> how can I get the invoiced timesheets of a created invoice to process outside Kimai?

See above: use one of the invoice events. 
You can't do it afterward and I am not interested in building an API  for that. 
But you can create some ManyToOne Entities in your bundle and store that mapping if you need it.

**Comment by DieserMerlin** [comment]
It would be very useful for us as well to query the timesheet items that an invoice covers. I think this might be a use case for a lot of users trying to integrate kimai with their accounting

**Reply by DieserMerlin** [reply]
I think the easiest way to implement that would be to set some meta data on the timesheet records when they appear in a specific invoice. This way you can also easily mark them as unexported after cancelling an invoice. Having to reset them manually doesn't seem too clean to me

**Comment by kernie** [comment]
The disadvantage of adding a field to the timesheet that stores the invoice ID comes then with a canceled invoice. The field would be cleared, however the canceled invoice exists further, has an invoice number and a grand total... 
I would prefer a combination: Add a boolean field `invoiced` to the timesheet to separate cleanly from the `exported`  state. Have any invoice store an array of the timesheet IDs. In case of the cancellation, only the field `invoiced` is reset.

**Reply by DieserMerlin** [reply]
Sounds pretty good to me, I wonder if the devs feel the same way?

**Reply by kevinpapst** [reply]
I tried once and ran into too many issues.The only idea that I would approve is: adding a JSON column that holds the type (timesheet, expense) and id of the included objects. That would allow to cancel the invoice.

I would keep the existing logic around the exported flag, as you have no chance of detecting whether an item was included in one or more invoices or exported otherwise. But we could include a confirmation dialog whether the included items should be set to not exported.

**Reply by kernie** [reply]
Thanks for the reply and sorry I have to ask... JSON column is how it is stored in the database, right? Basically:
```
{ "type": "timesheet", "timesheet_ids": [1,2,3,4,5] }
```
One query would result in the retrieval of the complete JSON object? 
Sounds like a good solution!

**Reply by kevinpapst** [reply]
Yes, we could even store it in a private meta-field, so we don't need to create a new column for it, as this data usually doesn't need to be read. I would probably vote for a structure like this: `[{type: "timesheet", entities: [1]}, {type: "expense", entities: [14]}]` because it allows to store the separate invoice rows (if grouping is used) and allows to be extended later on with additional attributes.

**Reply by DieserMerlin** [reply]
That honestly sounds pretty amazing, I was thinking about making a plugin for our use case but if this makes it into Kimai itself, I think a lot of people can profit from it

**Reply by kevinpapst** [reply]
> but if this makes it into Kimai itself, I think a lot of people can profit from it

I agree. Now someone  just needs to build it.
